Update README with new contribution guidelines

marcobonici opened the pull request:
Hello, I have
- included a preliminary version of the contributing guidelines
- fixed a couple of broken link, as found from @itutusaus 

Can someone give a look and merge into the main? Maybe @gcanasherrera ?

marcobonici commented:
Ehi @sfarrens , would you mind having a look at this?

sfarrens commented:
I will be happy to. 🙂

sfarrens commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+# Contributor's Guide on Collaborative Practices for cloe-org repositories
+
+This document describes some best practices for collaborating on the cloe-org repositories.
+Following these practices makes it easier for contributors (new and old) to understand what is expected of them.
+It should be linked to in the README.md. Those are meant to be general guidelines; it is possible that some repositories have additional guidelines.
+
+There are many good practices that this document does not cover.
+These include other members of the wider community reviewing pull requests (PRs) they are interested in, and maintainers encouraging and supporting people who open issues to make PRs to solve them.
+This document facilitates these other good practices by clarifying what can seem like a mysterious process to those who are unfamiliar with it.
```
I feel like the wording in general could use some work, but particularly in this section. I am happy to go into detail, but I am pretty sure even a quick pass with e.g. ChatGPT would clear this up.

sfarrens commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+    *   This makes it easier and more likely for all the smaller changes to be made.
+*   Reviewers should continue acting as reviewers until the PR is merged.
+
+### Accidental breaking releases
```
I guess this info is exclusive to repo maintainers, right?

marcobonici replied:
You mean, the Accidental breaking release?

marcobonici replied:
Answered as asked by Sam.

sfarrens replied:
Yes, the point is that general developers should not be the ones dealing with this type of issue.

marcobonici replied:
Added this information.

sfarrens commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+Hopefully, I can also add a test to prevent that part of the API from being broken by mistake.
+
+Now I look at my change again.
+If I can add the same functionality in a non-breaking way - for example, make a new internal function for my use - then I would do so and tag v1.15.2 or v1.16.0 depending on what had to change.Bu
```
```suggestion
If I can add the same functionality in a non-breaking way - for example, make a new internal function for my use - then I would do so and tag v1.15.2 or v1.16.0 depending on what had to change.
```

sfarrens reviewed (commented):
Hi @marcobonici, here are my initial comments on the contribution guidelines. Really nice work! 👏 I think just a bit of effort is needed to clean up the wording and make sure the roles/responsibilities are clear.

marcobonici commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+
+#### Contributing PRs
+
+*   PRs should match the existing code style present in the file.
```
So,
these are meant to be the general guidelines for the cloe-org. It can (and will) happen that each repository has specific guidelines.

marcobonici replied:
This has been written at the very beginning of the text, so that it's clear.

marcobonici commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+*   Comments and (possibly internal) docstrings should make the code accessible.
+*   PRs that change code must have appropriate tests.
+*   Changes to the code must be made via PR, not pushing to master.
+*   Before starting to work on a PR, especially when including or modifying a huge chunk of code, discuss with the mantainers the change you are gonna implement and how.
```
So, this is something to discuss with the others (@gcanasherrera @PedroCarrilho @chiaramoretti @AndreaPezzotta ).
For the moment, mantainers are in charge of the whole organization. Something different might happen in the future!

marcobonici commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+
+*   A release should be made as soon as possible after a bugfix PR is merged.
+*   Care and consideration should be given as to when to make a breaking release.
+*   The person who merged the PR should register the new release of the package.
```
@sfarrens , personally I would add a link to [semver](https://semver.org/). Any suggestion on something different?

marcobonici replied:
Added link to semver.

marcobonici commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+*   PRs with large improvements to style should not also change functionality.
+
+*   PRs introducing breaking changes should make this clear when opening the PR.
+*   You should not push commits with commented-out tests.
```
I modified according to the suggestion

marcobonici commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+
+There are many good practices that this document does not cover.
+These include other members of the wider community reviewing pull requests (PRs) they are interested in, and maintainers encouraging and supporting people who open issues to make PRs to solve them.
+This document facilitates these other good practices by clarifying what can seem like a mysterious process to those who are unfamiliar with it.
```
Rephrased, using perplexity.

marcobonici commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+
+*   PRs should match the existing code style present in the file.
+*   PRs affecting the public API, including adding new features, must update the public documentation.
+*   Comments and (possibly internal) docstrings should make the code accessible.
```
Modified as discussed with Sam.

marcobonici commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+*   PRs should pass CI tests before being merged.
+*   PRs by people __without__ merge rights must have approval from someone who has merge rights (who will usually then merge the PR).
+*   PRs by people __with__ merge rights must have approval from someone else, who may or may not have merge rights (and then may merge their own PR).
+*   PRs by people with merge rights should not be merged by people other than the author (just approved).
```
@sfarrens I think I simplified it a bit and made it more explicit. Also added a general background on dealing with PRs.

marcobonici commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+*   Small review suggestions, such as typo fixes, should make use of the `suggested change` feature.
+    *   This makes it easier and more likely for all the smaller changes to be made.
+*   Reviewers should continue acting as reviewers until the PR is merged.
+
```
Done as suggested.

marcobonici commented:
Hi @sfarrens , I think I addressed all of the points. There are still a bunch of open points, where I asked for feedback/suggestions.
Happy to go through the second round!

sfarrens commented on colprac/README.md:
```diff
@@ -0,0 +1,104 @@
+*   PRs affecting the public API, including adding new features, must update the public documentation.
+*   Comments and (possibly internal) docstrings should make the code accessible.
+*   PRs that change code must have appropriate tests.
+*   Changes to the code must be made via PR, not pushing to master.
```
Not exactly, depending on how many people are contributing to a package and what the release strategy is, it is not unreasonable for PRs to target `main`. My point is that no one should be 'pushing' to `main` (or `develop`), all changes should come from PRs that 'merge' into the appropriate branch. I would propose that you explicitly protect these branch so that no one can push to them.

marcobonici replied:
The rule ensures that the only way to commit to main is by PRs. Or am I missing something?

marcobonici commented:
Hi @sfarrens , I think I did everything.
I addressed your comments and passed chunks of the text to perplexity, that rephrased it with o3-mini.

Let me know whether there is something missing.

sfarrens commented:
@marcobonici 👍 to merge!
